tools/python/boutiques/localExec.py

```diff
@@ -392,60 +394,90 @@ def prepare(self):
-            # the image
-            lockfile = conName + ".lock"
-            lock = FileLock(lockfile)
-            lock.acquire()
+            return (conName, container_location)
+
+        if conType == 'singularity':
+            if not conIndex:
+                conIndex = "shub://"
+            elif not conIndex.endswith("://"):
+                conIndex = conIndex + "://"
+            conName = conImage.replace("/", "-").replace(":", "-") + ".simg"
+
+            # Check if container already exists
+            if self._singConExists(conName):
+                return (conName, "Local ({0})".format(conName))
+
+            # Container does not exist, try to pull it
+            lockdir = conName + "-lock"
+            maxcount = 90  # how long to wait before timing out
+            count = 0
             try:
-                if not conIndex:
-                    conIndex = "shub://"
-                elif not conIndex.endswith("://"):
-                    conIndex = conIndex + "://"
-                conName = conImage.replace("/", "-").replace(":", "-") + ".simg"
-
-                imagePath = './'
-                if self.imagePath is not None:
-                    imagePath = self.imagePath
-                    os.environ["SINGULARITY_PULLFOLDER"] = imagePath
-
-                if conName not in os.listdir(imagePath):
-                    pull_loc = "\"{0}\" {1}{2}".format(conName,
-                                                       conIndex,
-                                                       conImage)
-                    container_location = ("Pulled from {1}{2} ({0} not found "
-                                          "in current working "
-                                          "directory or specified "
-                                          "image path)").format(conName,
-                                                                conIndex,
-                                                                conImage)
-                    # Pull the singularity image
-                    sing_command = "singularity pull --name " + pull_loc
-                    (stdout, stderr), return_code = self._localExecute(
-                                                            sing_command)
-                    if return_code:
-                        message = ("Could not pull Singularity"
-                                   " image: " + os.linesep + " * Pull command: "
-                                   + sing_command + os.linesep + " * Error: "
-                                   + stderr.decode("utf-8"))
-                        raise ExecutorError(message)
-                else:
-                    container_location = "Local ({0})".format(conName)
+                while count < maxcount:
+                    count += 1
+                    try:
+                        os.mkdir(lockdir)
+                    except OSError:
+                        time.sleep(20)
```

Comment:
maybe reduce this time and increase maxcount: 360*5 instead of 90*20. The sleep time is the max delay that can be lost due to concurrent locks.